### PR TITLE
[SYCL-MLIR][MathOps] Model scalar math functions in `sycl` dialect

### PR DESCRIPTION
**Pull request by jopperm**
This PR models a subset† of the [SYCL math functions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_math_functions) as operations in the `sycl` dialect. Frontend support and mechanical 1:1 conversion patterns to `math` are also included.

† = ops that a have counterpart in the `math` dialect, and can be lowered via **`MathToLLVM`**. We considered lowering via the `spirv` dialect, however that doesn't fit well into the one-shot approach in `PolygeistToLLVM`, and doesn't add a meaningful number of supported math ops.

### CODE REVIEW

File: polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+void host_math_funcs(std::array<float, 1> &A) {
+  auto q = queue{};
+  device d = q.get_device();
+  std::cout << "Using " << d.get_info<info::device::name>() << "\n";
+
+  {
+    auto buf = buffer<float, 1>{A.data(), 1};
+    q.submit([&](handler &cgh) {
+      auto A = buf.get_access<access::mode::read_write>(cgh);
+      cgh.single_task<class kernel_math_funcs>([=]() {
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp:34]: We might not need a full program here. Maybe we can have:

```cpp
SYCL_EXTERNAL float f(float a, float b, float c) {
   ...
}
```

File: polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
+// RUN: clang++ -Xcgeist --use-opaque-pointers=1 -fsycl -fsycl-targets=spir64-unknown-unknown-syclmlir -O0 -w %s -o %t.O0.out 2>&1 | FileCheck %s --allow-empty --implicit-check-not="{{error|Error}}:"
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp:1]: I'd also like to have tests to check that the `sycl` operations are being generated in the first place (using `MLIR` as FileCheck prefix is the standard here)

**Review by victor-eds — COMMENTED** [review]
Can we get some tests for `double` too?

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+include "mlir/Interfaces/InferTypeOpInterface.td"
+
+////////////////////////////////////////////////////////////////////////////////
+// BASE DECLARATIONS
+////////////////////////////////////////////////////////////////////////////////
+
+// Generic floating-point types (§4.17.1).
+// TODO: Add vectors and marrays.
+def SYCLGenfloat : AnyTypeOf<[F32, F64]>;
```
[the remark this answers is not in the document]
Comment (jopperm) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td:24]: I'll go for the TODO here and address half support in a follow-up PR.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  "Compute `e^x - 1.0`">;
+def SYCLFabsOp : SYCL_FloatUnaryOp<"fabs",
+  "Compute absolute value of a floating-point number">;
+def SYCLFloorOp : SYCL_FloatUnaryOp<"floor",
+  "Round to integral value using the round to negative infinity rounding mode">;
+def SYCLFmaOp : SYCL_FloatTernaryOp<"fma"> {
+  let summary = "Computes fused multiply-add.";
+  let description = [{
+    Returns the correctly rounded floating-point representation of the sum of
+    `c` with the infinitely precise product of `a` and `b`.
```
[the remark this answers is not in the document]
Comment (jopperm) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td:84]: I see your point. My approach here was to follow the naming scheme (and the description) directly from the SYCL spec. In the rendered markdown docs, this looks much better:

-----

### `sycl.math.fma` (::mlir::sycl::SYCLFmaOp)

Computes fused multiply-add.


Syntax:

```
operation ::= `sycl.math.fma` $a `,` $b `,` $c attr-dict `:` type($result)
```

Returns the correctly rounded floating-point representation of the sum of
`c` with the infinitely precise product of `a` and `b`.

-----

WDYT?

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+struct FloatTernaryOpPattern : public OpConversionPattern<SYCLOpT> {
+  using OpConversionPattern<SYCLOpT>::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(SYCLOpT op, typename SYCLOpT::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<MathOpT>(op, adaptor.getA(), adaptor.getB(),
+                                         adaptor.getC());
+    return success();
+  }
+};
```
[the remark this answers is not in the document]
Comment (jopperm) [reply mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:64]: Absolutely, I'll change that. Thanks!

File: polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp (hunk 1 above)
Comment (jopperm) [reply polygeist/tools/cgeist/Test/Verification/sycl/math_funcs.cpp:34]: Yeah makes sense, I'll look into that.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td (hunk 4 above)
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td:84]: Yes, that LGTM. Thanks for the clarification!

**Review by victor-eds — COMMENTED** [review]
Can we also have a couple of examples in "invalid.mlir"?

**Review by victor-eds — APPROVED** [review]
All good to me. For the `invalid.mlir` test, I was thinking on a test in which not all of the types match, but it's true that would imply testing a core MLIR feature instead of our own verification. I leave keeping the current test up to you. Nice job!

**Review by whitneywhtsang — COMMENTED** [review]
What's the motivation to have math functions in `sycl` dialect instead of directly using the functions in `math` dialect?

**Comment by jopperm** [comment]
> What's the motivation to have math functions in `sycl` dialect instead of directly using the functions in `math` dialect?

We plan to uniformly support the SYCL-specific types (half, vectors, marray) and provide ops for math functions that currently don't have a counterpart in the upstream dialects.

**Review by whitneywhtsang — APPROVED** [review]

**Review by etiotto — APPROVED** [review]
LGTM left a couple of comments inline. Thanks for your first contribution to SYCL-MLIR!

**Pull request merged**